### ISSUE TITLE
Allow selection of groups as well as options (group members)

### ISSUE THREAD

**Issue by eddiesholl** [cibernox/ember-power-select#270]
Just wondering if there's any way to make group items selectable just like their option members. The grouping behaviour works perfectly for my scenario, but I just want to allow selection of groups, not just leaf nodes.

I think the only solution will be to simulate the grouping behaviour from scratch, to give myself the extra control.

**Comment by cibernox** [comment]
You're right. 

You cannot select groups, and that was an intentional decision.

For the record, I considered that at the beginning and even started an spike to implement it 
but abandoned it because it was a true nightmare to implement and not let that leak in the 
public API somehow.

It is possible with a little imagination but you need to mangle your options a bit.

My sugguestion is add before each group an option named as the group, and in the real group
not render the name, so the impression would be the same.

If your options are:

```
[
  { name: "one" },
  { name: "two" },
  { 
    groupName: "tens", 
    options: [
      { name: "twenty" },
      { name: "forty" }
    ]
  },
  { 
    groupName: "hundreds", 
    options: [
      { name: "three hundred" },
      { name: "five hundred" }
    ]
  }
```

you need to transform them in something like

```
[
  { name: "one" },
  { name: "two" },
  { name: "tens" },
  { 
    groupName: "tens", 
    options: [
      { name: "twenty" },
      { name: "forty" }
    ]
  },
  { name: "hundreds" },
  { 
    groupName: "hundreds", 
    options: [
      { name: "three hundred" },
      { name: "five hundred" }
    ]
  }
```

And pass a `optionsComponent` that behaves like the origina one but the template is different
and doesn't render then name of the group:

```hbs
{{#if loading}}
  {{#if loadingMessage}}
    <li class="ember-power-select-option" role="option">{{loadingMessage}}</li>
  {{/if}}
{{/if}}
{{#each options as |opt|}}
  {{#if opt.groupName}}
    <li class="ember-power-select-group" role="option">
      <!-- You don't want this line <span class="ember-power-select-group-name">{{opt.groupName}}</span> --> 
      {{#component optionsComponent highlighted=(readonly highlighted) selected=(readonly selected)
        options=(readonly opt.options) allOptions=(readonly allOptions) optionsComponent=(readonly optionsComponent) select=(readonly select)
        class="ember-power-select-options ember-power-select-options--nested" as |option|}}
        {{yield option lastSearchedText}}
      {{/component}}
    </li>
  {{else}}
    <li class="ember-power-select-option {{ember-power-select-option-classes opt selected highlighted}}"
      onmouseup={{action select.actions.choose (ember-power-select-build-selection opt selected multiple=multiple)}}
      onmouseover={{action select.actions.highlight opt}} role="option">
      {{yield opt lastSearchedText}}
    </li>
  {{/if}}
{{/each}}
```

That is going to give the sensation that the group's name is selectable, but in reality is just one option.

Perhaps when you select this fake group you want to set it as selected (so it gets highlighted properly) but also do something with the real one. You can get the original group finding the index of the fake one and incrementing it in one.

```hbs
{{#power-select options=options selected=selected onchange=(action "updateSelected") as |opt|}}
  {{opt}}
{{/power-select}}
```

```js
actions: {
  updateSelected(optionOrFakeGroup) {
    this.set('selected', selected);
    if (optionOrFakeGroup.__fake_group) {
      let index = this.get('options').indexOf(optionOrFakeGroup);
      let realGroup = this.get('options').objectAt(index + 1); // The group is the next option
      // Do something with it.
    } 
  }
}
```

Probably you can generalize this solution into an addon built on top of Ember Power Select,
like https://github.com/cibernox/ember-power-select-with-create. It also adds a fake option with
some flag to distinguish it.

If you want to do so, let me know, I'd be happy to help.

**Comment by eddiesholl** [comment]
Thanks for your help here. The solution I came up with could be bundled into a tree addon for power-select. Here's a run down of my approach. I hacked on the standard list format rather than groups and options:

* Each row has a depth field
* Render some non breaking spaces based on the depth of each row
* Use an optionsComponent that accepts both the option and the lastSearchedText
* Inside the optionsComponent, if the term matches, split the text into before, match and after match. Render the match part in bold (simulates typeahead).

I've also implemented some custom searching so that parent rows are included if their children match the search term. But there won't be any bold text in these items. And they are just regular rows, so selectable.

I'll post some code if I get a chance to do it, although some of my solution might not be super easy to generalise
